packages/syrup/src/encode.js

```diff
@@ -0,0 +1,227 @@
+function encodeRecord(buffer, record) {
+  const restart = buffer.length;
+  const indexes = [];
+  const keyStrings = [];
+  const keyBytes = [];
+
+  for (const key of keys(record)) {
+    const start = buffer.length;
+    encodeString(buffer, key);
+    const end = buffer.length;
+
+    keyStrings.push(key);
+    keyBytes.push(buffer.bytes.subarray(start, end));
+    indexes.push(indexes.length);
+  }
+
+  indexes.sort((i, j) =>
+    compareByteArrays(
+      keyBytes[i],
+      keyBytes[j],
+      0,
+      keyBytes[i].length,
+      0,
+      keyBytes[j].length,
+    ),
+  );
+
+  buffer.length = restart;
+
+  let cursor = grow(buffer, 1);
+  buffer.bytes[cursor] = DICT_START;
+
+  for (const index of indexes) {
+    const key = keyStrings[index];
+    const value = record[key];
+
+    encodeString(buffer, key);
+    // Recursion, it's a thing!
+    // eslint-disable-next-line no-use-before-define
+    encodeAny(buffer, value);
+  }
+
+  cursor = grow(buffer, 1);
+  buffer.bytes[cursor] = DICT_END;
+}
+
+/**
+ * @param {Buffer} buffer
+ * @param {Array<any>} array
+ */
+function encodeArray(buffer, array) {
+  let cursor = grow(buffer, 1);
+  buffer.bytes[cursor] = LIST_START;
+
+  for (const value of array) {
+    // Recursion, it's a thing!
+    // eslint-disable-next-line no-use-before-define
+    encodeAny(buffer, value);
+  }
+
+  cursor = grow(buffer, 1);
+  buffer.bytes[cursor] = LIST_END;
+}
+
+/**
+ * @param {Buffer} buffer
+ * @param {any} value
+ */
+function encodeAny(buffer, value) {
+  if (typeof value === 'string') {
+    encodeString(buffer, value);
+    return;
+  }
+
+  if (typeof value === 'number') {
+    const cursor = grow(buffer, 9);
+    buffer.bytes[cursor] = DOUBLE;
+    if (value === 0) {
+      // no-op
+    } else if (Number.isNaN(value)) {
+      // Canonicalize NaN
+      buffer.bytes.set(NAN64, cursor + 1);
+    } else {
+      buffer.data.setFloat64(cursor + 1, value, false); // big end
+    }
+    return;
+  }
+
+  if (typeof value === 'bigint') {
+    const string = value >= 0 ? `${value}+` : `${-value}-`;
+    const cursor = grow(buffer, string.length);
+    textEncoder.encodeInto(string, buffer.bytes.subarray(cursor));
+    return;
+  }
+
+  if (value instanceof Uint8Array) {
+    const prefix = `${value.length}:`; // decimal and colon suffix
+    const cursor = grow(buffer, prefix.length + value.length);
+    textEncoder.encodeInto(prefix, buffer.bytes.subarray(cursor));
+    buffer.bytes.set(value, cursor + prefix.length);
+    return;
+  }
+
+  if (Array.isArray(value)) {
+    encodeArray(buffer, value);
+    return;
+  }
+
+  if (Object(value) === value) {
+    encodeRecord(buffer, value);
+    return;
+  }
+
+  if (value === false) {
+    const cursor = grow(buffer, 1);
+    buffer.bytes[cursor] = FALSE;
+    return;
+  }
+
+  if (value === true) {
+    const cursor = grow(buffer, 1);
+    buffer.bytes[cursor] = TRUE;
+    return;
+  }
+
+  throw new TypeError(`Cannot syrialize value ${value}`);
```

Comment:
Are we syriously going to spell it this way?

Reply:
Only if no-one notices.